### PR TITLE
Fix Android openAuth async issues (issue #36)

### PR DESCRIPTION
**Pull request by Luke-Rogerson**
This is to address an issue with the `_openAuthSessionPolyfillAsync` method for Android. See #36 for details.

It ensures that the `_waitForRedirectAsync` method is properly awaited, the InAppBrowser is correctly opened and improves the promise resolution check for the `_redirectHandler`.

### CODE REVIEW

**Comment by viewpnt1** [comment]
if an error happens during login flow, _waitForRedirectAsync will never get resolved because no redirect happened server side, or am I wrong?

**Comment by Luke-Rogerson** [comment]
@viewpnt1 Sometimes, `_waitForRedirectAsync` is getting immediately resolved as `type: success`, which we don't want, as at that point we don't have the `auth_code` appended to the end of the URL, which is why I added that additional check.

However, perhaps there is something I'm not fully understanding. @jdnichollsc could you help us here and explain a bit more about when `_waitForRedirectAsync` should be resolved? Is the check I added for a param not the correct way to properly wait for a valid response?

**Comment by Luke-Rogerson** [comment]
@viewpnt1 Sometimes, `_waitForRedirectAsync` is getting immediately resolved as `type: success`, which we don't want, as at that point we don't have the `auth_code` appended to the end of the URL, which is why I added that additional check.

However, perhaps there is something I'm not fully understanding. @jdnichollsc could you help us here and explain a bit more about when `_waitForRedirectAsync` should be resolved? Is the check I added for a param not the correct way to properly wait for a valid response?

File: index.js [hunk 1]
```diff
@@ -84,7 +84,10 @@ async function open(
       inAppBrowseroptions.preferredControlTintColor
     );
   }
-  return RNInAppBrowser.open(inAppBrowseroptions);
+  // use SetTimeout to make sure InAppBrowser doesn't open until current execution context has finished
+  setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (Luke-Rogerson) [reply index.js]: Do you have any alternatives then? I know it's not a particularly elegant solution, but I was getting very hit or miss results when the in app browser modal was actually opening. Adding the `setTimeout` resolved this.

File: index.js [hunk 2]
```diff
@@ -130,32 +133,33 @@ async function _openAuthSessionPolyfillAsync(
   returnUrl: string,
   options: InAppBrowserOptions
 ): Promise<AuthSessionResult> {
+  // reset redirect handler upon new Auth session
+  _redirectHandler = null
```
[the remark this answers is not in the document]
Comment (Luke-Rogerson) [reply index.js]: OK, fair enough

File: index.js [hunk 3]
```diff
@@ -130,32 +133,33 @@ async function _openAuthSessionPolyfillAsync(
...
 }
 
-function _waitForRedirectAsync(returnUrl: string): Promise<RedirectResult> {
+async function _waitForRedirectAsync(returnUrl: string): Promise<RedirectResult> {
   return new Promise(resolve => {
-    _redirectHandler = (event: RedirectEvent) => {
-      if (event.url.startsWith(returnUrl)) {
+    _redirectHandler = (event: RedirectEvent) => {      
+       // check URL has param before resolving (eg. example://auth?code=abcd) 
+          if (event.url.startsWith(returnUrl + '?')) {
```
[the remark this answers is not in the document]
Comment (Luke-Rogerson) [reply index.js]: Is it not? Are we not always expecting the redirect uri + a `code` param as the resolved value from this function?

File: index.js [hunk 4]
```diff
@@ -130,32 +133,33 @@ async function _openAuthSessionPolyfillAsync(
...
     !_redirectHandler,
     'InAppBrowser.openAuth is in a bad state. _redirectHandler is defined when it should not be.'
   );
 
   try {
-    return await Promise.race([
-      open(startUrl, options),
-      _waitForRedirectAsync(returnUrl)
-    ]);
+    await open(startUrl, options)
```
[the remark this answers is not in the document]
Comment (Luke-Rogerson) [reply index.js]: What kind of delay were you thinking? 1000ms?

**Comment by Luke-Rogerson** [comment]
> Check my comments and let me know 👍

Thanks for the review! Please see my responses 😎

File: index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdnichollsc) [reply index.js]: Can you share your code? we don't have any message :)

File: index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdnichollsc) [reply index.js]: Params are optional 👍

File: index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdnichollsc) [reply index.js]: I mean adding a new option to configure that

**Comment by jdnichollsc** [comment]
Thanks guys for your help, we're going to use this other PR instead https://github.com/proyecto26/react-native-inappbrowser/pull/103

**Comment by Luke-Rogerson** [comment]
No problem. Apologies I wasn't able to update this PR. I've been away from my computer for the last week or so!
Thanks to you for all your work on this library.

**Comment by jdnichollsc** [comment]
No problem bro, I want to fix that issue too

**Pull request closed without merging**